### PR TITLE
Make `diff.namespace.to-patch` use "v2" branch diffs

### LINKED ISSUE #3683: Switch diff.namespace.to-patch to use "v2" namespace diff
Discussion from #3646: https://github.com/unisonweb/unison/pull/3646#discussion_r1042489199

### LINKED ISSUE #3685: Handle aliases better in `diff.namespace.to-patch`
Discussion: https://github.com/unisonweb/unison/pull/3646#discussion_r1042514162

### PR DESCRIPTION
**Pull request by mitchellwrosen**
- [x] Depends on #3691
- [x] Depends on #3692

---

## Overview

Fixes #3683 
Fixes #3685 

This PR moves `diff.namespace.to-patch` over to a more efficient implementation based on "v2 branch diffs", which don't leverage the in-memory branches.

### CODE REVIEW

File: codebase2/codebase/U/Codebase/Referent.hs [hunk 1]
```diff
@@ -54,6 +54,11 @@ toReference = \case
   Ref termRef -> termRef
   Con typeRef _ -> typeRef
 
+toTermReference :: Referent' termRef typeRef -> Maybe termRef
```
Comment (ChrisPenner) [inline codebase2/codebase/U/Codebase/Referent.hs:57]: This behaves differently than `Referent.toReference` which I think is likely to be a footgun in the future;
https://github.com/unisonweb/unison/blob/trunk/unison-core/src/Unison/Referent'.hs#L57-L60

Not saying the behaviour of `Referent.toReference` is necessarily correct, but we definitely depend on its current behaviour a lot in  the `Backend` module right now 😬 

The fact that this returns a Maybe is hopefully enough of a signifier about what it's doing, but it still makes me a little nervous to have differing behaviour there.

File: parser-typechecker/src/U/Codebase/Branch/Diff.hs [hunk 2]
```diff
@@ -83,56 +88,71 @@ data NameChanges = NameChanges
...
+--   2. `a` != `b`
+data NameBasedDiff = NameBasedDiff
+  { terms :: Relation Reference Reference,
+    types :: Relation Reference Reference
+  }
+  deriving stock (Generic, Show)
+
+instance Monoid NameBasedDiff where
+  mempty = NameBasedDiff mempty mempty
+  mappend = (<>)
```
Comment (ChrisPenner) [inline parser-typechecker/src/U/Codebase/Branch/Diff.hs]: I don't think we need `mappend` defined anymore, in fact I think some GHC's might even complain about it if you do? 🤔

File: parser-typechecker/src/U/Codebase/Branch/Diff.hs [hunk 3]
```diff
@@ -172,3 +189,27 @@ nameChanges namePrefix (TreeDiff (DefinitionDiffs {termDiffs, typeDiffs} :< chil
       xs
         & Set.toList
         & fmap (name,)
+
+-- | Get a 'NameBasedDiff' from a 'TreeDiff'.
+nameBasedDiff :: TreeDiff -> NameBasedDiff
+nameBasedDiff (TreeDiff (DefinitionDiffs {termDiffs, typeDiffs} :< children)) =
```
Comment (ChrisPenner) [inline parser-typechecker/src/U/Codebase/Branch/Diff.hs:195]: TreeDiff just wraps a Cofree, so I believe you can factor out the recursion into a single foldMap here if you like.

```suggestion
nameBasedDiff (TreeDiff d) =
  d & foldMap \defDiff -> 
        NameBasedDiff 
        { terms = nameBasedTermDiff defDiff
        , types = nameBasedTypeDiff defDiff
        }
```
                
```

File: parser-typechecker/src/U/Codebase/Branch/Diff.hs [hunk 4]
```diff
@@ -172,3 +189,27 @@ nameChanges namePrefix (TreeDiff (DefinitionDiffs {termDiffs, typeDiffs} :< chil
...
+  let NameBasedDiff childrenTerms childrenTypes =
+        foldMap (nameBasedDiff . TreeDiff) children
+   in NameBasedDiff
+        { terms = foldMap nameBasedTermDiff termDiffs <> childrenTerms,
+          types = foldMap nameBasedTypeDiff typeDiffs <> childrenTypes
+        }
```
Comment (ChrisPenner) [inline parser-typechecker/src/U/Codebase/Branch/Diff.hs]: TreeDiff just wraps a Cofree, so I believe you can factor out the recursion into a single foldMap here if you like.

```suggestion
nameBasedDiff (TreeDiff d) =
  d & foldMap \DefinitionDiffs{termDiffs, typeDiffs} ->
        NameBasedDiff
        { terms = foldMap nameBasedTermDiff termDiffs
        , types = foldMap nameBasedTypeDiff typeDiffs
        }
```

File: parser-typechecker/src/U/Codebase/Branch/Diff.hs [hunk 5]
```diff
@@ -172,3 +189,27 @@ nameChanges namePrefix (TreeDiff (DefinitionDiffs {termDiffs, typeDiffs} :< chil
...
+  where
+    nameBasedTermDiff :: Diff Referent -> Relation Reference Reference
+    nameBasedTermDiff Diff {adds, removals} =
+      let termAdds = mapMaybe Referent.toTermReference (Set.toList adds)
+          termRemovals = mapMaybe Referent.toTermReference (Set.toList removals)
```
Comment (ChrisPenner) [inline parser-typechecker/src/U/Codebase/Branch/Diff.hs:204]: Is there a particular reason we don't patch constructors?

I don't think that change would happen in this PR, but it seems like it'd fit naturally in here.

File: parser-typechecker/src/U/Codebase/Branch/Diff.hs (hunk 2 above)
Comment (mitchellwrosen) [reply parser-typechecker/src/U/Codebase/Branch/Diff.hs]: Oh, whoops!

File: parser-typechecker/src/U/Codebase/Branch/Diff.hs (hunk 5 above)
Comment (mitchellwrosen) [reply parser-typechecker/src/U/Codebase/Branch/Diff.hs:204]: They just aren't representable in a patch right now, so I left them out of this structure, too.

Old ticket: https://github.com/unisonweb/unison/issues/3504

File: unison-cli/src/Unison/Cli/MonadUtils.hs [hunk 6]
```diff
@@ -144,32 +148,54 @@ resolveAbsBranchId = \case
...
-        len <- Codebase.branchHashLength
-        pure (hashSet, len)
-    h <-
-      Set.asSingleton hashSet & onNothing do
-        Cli.returnEarly
-          if Set.null hashSet
-            then Output.NoBranchWithHash hash
-            else Output.BranchHashAmbiguous hash (Set.map (SCH.fromHash len) hashSet)
-    branch <- liftIO (Codebase.getBranchForHash codebase h)
+    hash <- Cli.runTransaction (resolveShortCausalHashToCausalHash shortHash) & onLeftM Cli.returnEarly
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Cli/MonadUtils.hs]:
```suggestion
    hash <- Cli.runEitherTransaction (resolveShortCausalHashToCausalHash shortHash)
```

File: parser-typechecker/src/U/Codebase/Branch/Diff.hs (hunk 4 above)
Comment (mitchellwrosen) [reply parser-typechecker/src/U/Codebase/Branch/Diff.hs]: I'm not sure I understand. The suggestion doesn't type check.

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 7]
```diff
@@ -1804,58 +1806,34 @@ handleDiffNamespaceToPatch :: Text -> DiffNamespaceToPatchInput -> Cli ()
...
-  let patch =
-        Patch
-          { _termEdits = Relation.fromSet termUpdates,
-            _typeEdits = Relation.fromSet typeUpdates
-          }
+  absBranchId1 <- Cli.resolveBranchIdToAbsBranchId (input ^. #branchId1)
+  absBranchId2 <- Cli.resolveBranchIdToAbsBranchId (input ^. #branchId2)
+
+  patch <- do
+    Cli.runEitherTransaction do
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1813]:
```suggestion
  patch <- Cli.runEitherTransaction do
```

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 8]
```diff
@@ -1804,58 +1806,34 @@ handleDiffNamespaceToPatch :: Text -> DiffNamespaceToPatchInput -> Cli ()
...
+      runExceptT do
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1814]:
```suggestion
  patch <- Cli.runEitherTransaction . runExceptT do
```

File: parser-typechecker/src/U/Codebase/Branch/Diff.hs (hunk 4 above)
Comment (mitchellwrosen) [reply parser-typechecker/src/U/Codebase/Branch/Diff.hs]: Oh, I think I get it.

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs (hunk 8 above)
Comment (mitchellwrosen) [reply unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1814]: Eh? I don't think this is clearer

File: parser-typechecker/src/U/Codebase/Branch/Diff.hs (hunk 4 above)
Comment (ChrisPenner) [reply parser-typechecker/src/U/Codebase/Branch/Diff.hs]: Oh, I missed a foldMap, fixed it.

It's not really any shorter, but personally I like to factor out recursion when I can, I find it helps with readability 😄

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs (hunk 8 above)
Comment (ChrisPenner) [reply unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1814]: Personal preference 🤷🏼‍♂️

**Review by ChrisPenner — APPROVED** [review]
Thanks for swapping over 👍🏼

**Pull request merged**